server/src/main/java/org/elasticsearch/search/SearchHit.java

```diff
@@ -724,6 +729,17 @@ private void deallocate() {
             r.decRef();
         }
         SearchHit.this.source = null;
+        clearIfMutable(documentFields);
+        clearIfMutable(metaFields);
+        this.highlightFields = null;
+    }
+
+    private static void clearIfMutable(Map<String, DocumentField> fields) {
+        // check that we're dealing with a HashMap, instances read from the wire that are empty be of an immutable type
+        assert fields instanceof HashMap<?, ?> || fields.isEmpty() : fields;
+        if (fields instanceof HashMap<?, ?> hm) {
```

Comment:
should we just check if the `fields` map is empty as opposed to an instanceof check? (note that `EmptyMap.clear()` doesn't throw, however a potential replacement with `List.of()` would)